FEATURE: User fields required for existing users - Part 2

Drenmi opened the pull request:
### What is this feature?

We want to allow admins to make new required fields apply to existing users. In order for this to work we need to have a way to make those users fill up the fields on their next page load. This is very similar to how adding a 2FA requirement post-fact works. Users will be redirected to a page where they can fill up the remaining required fields, and until they do that they won't be able to do anything else.

### Implementation details

As an optimization, we keep a version number of the current "set of required fields" in `UserRequiredFieldsVersion`. We check this against a `User#required_fields_version` field. If the user's version is out of date, we perform a check to see if they have filled up all required fields. If they have, we update their version. If they haven't, we redirect to the preferences profile page and show them only the required fields they need to fill up.

### Screenshots

**When admin is adding a new field:**

<img width="559" alt="Screenshot 2024-06-07 at 10 58 56 AM" src="https://github.com/discourse/discourse/assets/5259935/5b644a5a-a64c-47bd-a95b-6c73923f70ab">

**User getting asked to fill up fields:**

<img width="597" alt="Screenshot 2024-06-07 at 11 42 23 AM" src="https://github.com/discourse/discourse/assets/5259935/aa998f74-d75f-4ff7-b693-ccd186f1b293">

martin-brennan commented on app/assets/javascripts/discourse/app/controllers/preferences/profile.js:
```diff
@@ -81,6 +93,14 @@ export default Controller.extend({
   },
 
+  _missingRequiredFields(siteFields, userFields) {
+    return siteFields
```
Might actually be better to do this in the `user.js` model, you have access to `Site` from there, and this will help remove the dependency on the controller in the route file for the profile

Drenmi replied:
I guess the concern here is mainly having a route reference its controller? I assume this is less conventional since there's no "native" access, need to use `controllerFor`. At the same time I'm a bit hesitant to chuck this in `User` since it's UI logic related to this particular view only, so it 1) creates more indirection, 2) puts UI specific stuff in the view, and 3) adds to the 1.5k LOC already in `User`. 😬

martin-brennan replied:
Yes that's it, it feels very odd for the route have to dig into controller logic like this (also the fact that so much logic is in the controller since it's an old route is :nauseated_face: ), and also since the condition is `showEnforcedRequiredFieldsNotice` which is mostly about a UI element on the page.

I just realised that is only pointing to `user.needs_required_fields_check` , so we should check `user.needs_required_fields_check` directly in the route, that would work better for me.

Drenmi replied:
> I just realised that is only pointing to user.needs_required_fields_check , so we should check user.needs_required_fields_check directly in the route, that would work better for me.

This is it! 🚀

martin-brennan commented on app/assets/javascripts/discourse/app/routes/preferences-profile.js:
```diff
@@ -1,7 +1,24 @@
+    super.willTransition(...arguments);
+
+    if (
+      this.controllerFor("preferences.profile").get(
```
Hmm I would probably not rely on a controller property for this, see note about `_missingRequiredFields`, then you could do `model.missingRequiredFields.length > 0`.

I think also if admins are trying to get into an admin route we need to allow this...otherwise they can lock themselves and all other admins out of the admin routes.

martin-brennan replied:
I realised in my other comment this actually just points to `user.needs_required_fields_check`, so let's rely on this here instead of digging into the controller

martin-brennan reviewed (commented):
Okay a fair few comments here, LMK if you need any clarifications :) 

I think you need to add a `Discourser.request_refresh!` line to `Admin::UserFieldsController` in create/update calls. I ran into a problem where, as an admin where I set up a new required for all users field, I started getting errors when I navigated around the site until refresh. When you use `Discourser.request_refresh!` on next ember transition the whole window refreshes.

Confirming too that `enforce_second_factor` takes precedence for redirects?

I think we should only show the "for all users" required fields on the profile page when redirecting there -- I see 2 that are "on signup" fields (the top ones), I think this is confusing:

![image](https://github.com/discourse/discourse/assets/920448/eef92776-7000-4ed0-83df-7209db384d54)

Also, the "How to?" field is _not_ editable after signup, so I'm not sure why I can even change it.

Also, are you going to add these in a followup PR?

* Confirm dialog for admins “This will prompt COUNT users to fill in this field and will not allow them to do anything else on your site until the field is filled, proceed?”
* UserHistory logs for every time a user hits this page and also for when they fill out the fields
* Update the admin user field controller to use the new Service::Base system in core

martin-brennan commented:
I suspect the `SiteSetting.enforce_second_factor` is _not_ taking precedence, I currently get a "too many redirects" error when this setting is set to `staff` on a user which has not filled out all the required fields:

![image](https://github.com/discourse/discourse/assets/920448/b041d3f8-eed4-4dea-b6f0-e692cb9f68b5)

Flink commented on app/models/user_field.rb:
```diff
@@ -14,9 +14,12 @@ class UserField < ActiveRecord::Base
   after_save :queue_index_search
 
   scope :public_fields, -> { where(show_on_profile: true).or(where(show_on_user_card: true)) }
+  scope :required, -> { where.not(requirement: 0) }
```
Here we can just use the scope automatically created by the `enum`: `not_optional` :slightly_smiling_face:

Drenmi replied:
I forgot this was added to Rails. Thank you, @Flink! ☺️

Flink commented on app/serializers/current_user_serializer.rb:
```diff
@@ -288,6 +289,10 @@ def include_associated_account_ids?
     SiteSetting.include_associated_account_ids
   end
 
+  def needs_required_fields_check?
```
As we’re not processing the value of `#needs_required_fields_check?`, I don’t think we need this method. `attributes :needs_required_fields_check?` should already do the job if I’m not mistaken.

Drenmi replied:
![](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaWw4MjE3bTFjbnc5M3NtYWpqNmdoNW52djl5bG1pejQ3NHl0cHM5ayZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3oFzmkkwfOGlzZ0gxi/giphy.webp)

Flink commented on spec/system/user_page/user_preferences_profile_spec.rb:
```diff
@@ -0,0 +1,47 @@
+        field_type: "text",
+        name: "Favourite Pokemon",
+        description: "Hint: It's Mudkip.",
+        requirement: 1,
```
Since this is an AR enum, we should even be able to just do `requirement: :for_all_users`.

Drenmi replied:
![](https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExbjVmOWt6dG5vMmN2cDA4YzJzdWR4N2kxdzRlY2FuOTJuYWRjZm5nZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/jofMaHVXT8QVYWlHwJ/giphy.webp)

Drenmi commented on app/assets/javascripts/admin/addon/components/admin-user-field-item.hbs:
```diff
@@ -57,7 +76,11 @@
+        <Input
+          @type="checkbox"
+          @checked={{this.buffered.editable}}
+          disabled={{this.buffered.editable}}
```
Wow. Can't believe I missed this. Got a but myopic with what I was trying to do. 😂

Drenmi commented on app/controllers/application_controller.rb:
```diff
@@ -929,6 +930,23 @@ def should_enforce_2fa?
+    return if !current_user
+    return if !current_user.needs_required_fields_check?
+
+    if current_user.populated_required_custom_fields?
```
The crux here is we can't guarantee this will work because by design the version number and the fields might not agree. In particular if an admin creates a new required field (bumps the version) and then changes it to be not required (or deletes it). The user's version will now be out of date, but they have no missing required fields.

> Otherwise we have to check this every page load, when we already know if they need the check or not in a cheaper call on line 936 above.

The L:936 one is a short-circuit, so it's only every page load for a user with an out-of-date version (i.e. not in the hot path.)

martin-brennan replied:
Okay cool, sounds reasonable to me :+1:

Drenmi commented on app/controllers/application_controller.rb:
```diff
@@ -929,6 +930,23 @@ def should_enforce_2fa?
+    end
+
+    redirect_path = path("/u/#{current_user.encoded_username}/preferences/profile")
+    if !request.fullpath.start_with?(redirect_path)
```
I think we should address this in Part 3 when we bring 2FA and required fields focus modes together. (We currently have the exact same situation for 2FA requirement.) It's not really bricked since you can just input some dummy value to be free again.

martin-brennan replied:
Yes but it's still a very bad experience for admins, and they may be reluctant to input dud data here. They could also have tons of fields to fill in. I would really prefer we address this in the current PR.

martin-brennan replied:
Yes but it's still a very bad experience for admins, and they may be reluctant to input dud data here. They could also have tons of fields to fill in. I would really prefer we address this in the current PR.

Drenmi commented on app/assets/javascripts/discourse/app/controllers/preferences/profile.js:
```diff
@@ -53,6 +60,11 @@ export default Controller.extend({
   },
 
+  @discourseComputed("currentUser")
+  showEnforcedRequiredFieldsNotice(user) {
```
Ah. Good point. Agree. Probably confusing to just have the "saved" flash there in this case.

martin-brennan replied:
This is still happening after your changes. I think we need to update _something_ on the user model to let them navigate away from the page, I am guessing `user.needs_required_fields_check` since that's used in willTransition on the route.

Drenmi commented on db/migrate/20240531053226_create_user_required_fields_version.rb:
```diff
@@ -0,0 +1,11 @@
+      t.timestamps null: false
+    end
+
+    add_column :users, :required_fields_version, :integer
```
Yep. I went for a conditional in the application code instead for this exact reason.

Drenmi commented on spec/system/user_page/user_preferences_profile_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    before do
+      UserRequiredFieldsVersion.create!
+      UserField.create!(
+        field_type: "text",
```
I was very surprised as well. I'm worried this was intended to be extendable and maybe used in some plugin?

Drenmi replied:
Couldn't find any usages in `discourse` or `discourse-org`. Will add a TODO and separate PR for this so it doesn't risk us having to roll back the feature if it breaks something.

Drenmi commented on spec/system/user_page/user_preferences_profile_spec.rb:
```diff
@@ -0,0 +1,47 @@
+      expect(page).to have_selector(
+        ".alert-error",
+        text:
+          "You are required to provide additional information before continuing to use this site.",
```
Sure. This is just a copy-pasta from the 2FA test. 🙂

Drenmi commented on spec/system/user_page/user_preferences_profile_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+      visit("/")
+
+      expect(page).to have_text("You're all caught up!")
```
Nice! Was looking for something like an idiomatic way to do this. Do you think we should have a dedicated method for this?

martin-brennan commented:
@Drenmi I added some more comments, not sure you saw my main comments at https://github.com/discourse/discourse/pull/27172#pullrequestreview-2106647658 from yesterday too?

martin-brennan commented on spec/system/user_page/user_preferences_profile_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+      expect(page).to have_current_path("/u/bruce0/preferences/profile")
+
+      # Shows a notice that the user needs to fill up more fields.
```
Like I said in chat too I really don't think the comments are necessary and we should split the redirect test into another test. Thanks for doing the I18n stuff though!

Drenmi commented:
> I ran into a problem where, as an admin where I set up a new required for all users field, I started getting errors when I navigated around the site until refresh.

This should be fixed now that we allow admin routes even when the admin is missing required fields.

> Confirming too that enforce_second_factor takes precedence for redirects?

Confirmed fixed.

> I think we should only show the "for all users" required fields on the profile page when redirecting there -- I see 2 that are "on signup" fields (the top ones), I think this is confusing.

I see what you mean. Changed the condition to be "for all users" rather than "required".

> Also, the "How to?" field is not editable after signup, so I'm not sure why I can even change it.

Fixed as part of the above comment.

> Confirm dialog for admins “This will prompt COUNT users to fill in this field and will not allow them to do anything else on your site until the field is filled, proceed?”

Added a confirmation in this PR. I made it a static message without the dynamic "count". Looking into if we already have the count in our state.

> UserHistory logs for every time a user hits this page and also for when they fill out the fields

Added in this PR.

> Update the admin user field controller to use the new Service::Base system in core

This is the only one planned for the next PR.

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-user-field-item.js:
```diff
@@ -64,8 +66,29 @@ export default Component.extend(bufferedProperty("userField"), {
+    this.buffered.set("editable", requirement === "for_all_users");
+  },
+
+  async _confirmChanges() {
```
I did this wrong in my PR apparently :sweat_smile: I am trying to get clarification from cvx about it. For now this is fine :+1:

martin-brennan commented on app/assets/javascripts/discourse/app/routes/preferences-profile.js:
```diff
@@ -1,7 +1,23 @@
+  willTransition(transition) {
+    super.willTransition(...arguments);
+
+    if (this.currentUser.needs_required_fields_check) {
```
I think _maybe_ `currentUser` will be null here for anonymous users, try visiting someone's profile page as anon user to double check. Might be fine

martin-brennan replied:
We also need to allow transitioning to an admin route, checking `transition?.to.name.startsWith("admin")` should be fine

martin-brennan commented on app/controllers/application_controller.rb:
```diff
@@ -929,6 +930,29 @@ def should_enforce_2fa?
+
+    redirect_path = path("/u/#{current_user.encoded_username}/preferences/profile")
+    second_factor_path = path("/u/#{current_user.encoded_username}/preferences/second-factor")
+    allowed_paths = [redirect_path, second_factor_path, "/admin"]
```
Do we need to wrap `/admin` with `path()`?

Drenmi replied:
I suspect we might ... subfolder thingy?

martin-brennan commented on app/controllers/application_controller.rb:
```diff
@@ -929,6 +930,29 @@ def should_enforce_2fa?
+    second_factor_path = path("/u/#{current_user.encoded_username}/preferences/second-factor")
+    allowed_paths = [redirect_path, second_factor_path, "/admin"]
+    if allowed_paths.none? { |p| request.fullpath.start_with?(p) }
+      UserHistory.create!(
```
This only needs to be once per day, otherwise the user could end up making tons of these. This can be done in redis with a combination of `setnx` and `expire`, here is another example:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/models/topic_link_click.rb#L109-L114

If `setnx` returns true, it means the value did not exist https://github.com/redis/redis-rb/blob/433ec69c563f1b51dead336fc81f1fb70a8297af/lib/redis/commands/strings.rb#L125

It would be nice to eventually extract this pattern into a useful oneliner, like:

```ruby
Discourse.redis.perform_once_in_period(key, seconds) do
  // the thing
end
```

martin-brennan replied:
This and the 302 error on redirect from /admin back to the forum are the most important things to change before merging

martin-brennan replied:
Can you also please write a spec for it

Drenmi replied:
> It would be nice to eventually extract this pattern into a useful oneliner.

I think this could even go in as part of the `UserHistory` API in addition to having the low-level one-liner.

martin-brennan commented on app/services/user_updater.rb:
```diff
@@ -266,6 +266,13 @@ def update(attributes = {})
       end
       DiscourseEvent.trigger(:user_updated, user)
+
+      if attributes[:custom_fields].present? && user.needs_required_fields_check?
```
Can you write a spec to cover this please?

martin-brennan commented on app/controllers/application_controller.rb:
```diff
@@ -929,6 +930,29 @@ def should_enforce_2fa?
   end
 
+  def redirect_to_profile_if_required
+    return if request.format.json? && is_api?
```
I realised that we actually need to just `return if request.format.json?` here, if you take another look at `should_enforce_2fa?` we actually do not do the redirect if the request is JSON which makes sense:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/controllers/application_controller.rb#L922-L923

It's confusing because it's mixed up with the "redirect to login" logic:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/controllers/application_controller.rb#L878-L879

This is why it will be excellent when you refactor this so the whole "redirect user if we want to force them to do something" logic will be in the same place :)

For now we should just fix this case and add a request spec to catch it. The steps I followed to find the issue:

1. Make a required for all field
2. As an admin, visit /admin route manually in my browser address bar
3. Click the "Back to forum" button in the sidebar
4. Get errors in the topic list, because it cannot handle a 302 response from the server

At the point of 4, I think we actually need a top-level `willTransition` check in our application.js route:

```
@action
willTransition(transition) {
  if (this.currentUser.needs_required_fields_check && transition.to?.name != "preferences.profile") {
      transition.abort();
      return this.router.transitionTo("preferences.profile");
  }

  return true;
},
```

This will be helpful for the 2FA route, I am guessing a similar loophole exists there now.

Drenmi replied:
> At the point of 4, I think we actually need a top-level willTransition check in our application.js route

Yes. I have a separate issue opened for this relating to 2FA, where you can go to the `/faq` page and then use client-side routing from there. I have a fix coming up for that. It should cover this use case as well. 🙏

martin-brennan commented on app/assets/javascripts/admin/addon/components/admin-user-field-item.js:
```diff
@@ -79,6 +102,17 @@ export default Component.extend(bufferedProperty("userField"), {
+    }
+
+    if (!confirm) {
+      this.cancel();
```
I don't think we need this, it resets the whole form. We can just return, then they can make tweaks to the requirements and re-submit

martin-brennan reviewed (commented):
Sorry there is more to fix here. This redirection stuff is a :bear:

martin-brennan reviewed (approved):
Okay I think this is looking good, thanks for bearing with me :) I look forward to your top-level fix to the application route in Ember to clear up all the various redirect logic code :) Nice use of rate limiter for the once per day user history too :ok_hand: 

Approving but I think one last thing to add is this:

> I think you need to add a Discourser.request_refresh! line to Admin::UserFieldsController in create/update calls. When you use `Discourser.request_refresh!` on next Ember transition the whole window refreshes.

Otherwise as admin you can make the new field, leave the admin portal, and not be prompted to fill the field until you refresh again, and existing users already logged in and clicking around won't be prompted until they refresh again.
